### DISCUSSION TITLE
What's the easiest way to encode a String to utf-8?

### DISCUSSION THREAD

**Discussion by sa-** [modular/modular#110] [Q&A] [answered]
I have a String and I would like to convert it into a byte array (so presumably a `DynamicVector[UI8]`)

**Comment by czheo** [comment, marked as answer]
maybe this
``` python
from Vector import DynamicVector
from String import String

let vec : DynamicVector[SI8] = String("hello world").buffer
```

**Reply by sa-** [reply]
Nice, thanks!

**Comment by survuvi** [comment]
and how to print the vec variable, we need builtins module?

```python
from Vector import DynamicVector
from String import String
from PythonInterface import Python

let vec : DynamicVector[SI8] = String("hello world").buffer

let builtins = Python.import_module("builtins")
builtins.print(vec)
```

error: Expression [4]:23:19: invalid call to '__call__': argument #1 cannot be converted from 'DynamicVector[SIMD[si8, 1]]' to 'PythonObject'
    builtins.print(vec)
    ~~~~~~~~~~~~~~^~~~~

/.modular/Kernels/mojo/Python/PythonObject.mojo:210:5: function declared here
    fn __call__(self, *args: PythonObject) raises -> PythonObject:
    ^

**Comment by survuvi** [comment]
```python

from Vector import DynamicVector
from String import String

fn conversion_to_utf(my_string) -> str:
    let vec : DynamicVector[SI8] = String(my_string).buffer
    return vec
    
conversion_to_utf("hello world")

```

_error: Expression [13]:8:36: use of unknown declaration 'str', 'fn' declarations require explicit variable declarations
fn conversion_to_utf(my_string) -> str:
                                   ^~~

error: Expression [13]:8:22: 'fn' parameter type must be specified
fn conversion_to_utf(my_string) -> str:_

**Reply by abhinav-upadhyay** [reply]
You are returning `DynamicVector[SI8]` so you need to declare that as the return type of the function.
`fn conversion_to_utf(my_string) -> DynamicVector[SI8]:`

The other problem is that when writing Mojo functions (fn) -- you need to declare the types of all the arguments to the function. So it should be something like

`fn conversion_to_utf(my_string: String) -> DynamicVector[SI8]:`

Your final code will look like this:
```python
from Vector import DynamicVector
from String import String

fn conversion_to_utf(my_string: String) -> DynamicVector[SI8]:
    let vec : DynamicVector[SI8] = my_string.buffer
    return vec
    
conversion_to_utf("hello world")
```

**Reply by survuvi** [reply]
why it doesn't show any output value?

**Reply by mfranzon** [reply]
From my understanding, at the moment there is not a method to stored the object itself into the dynamic vector but just its reference. So, if you `print` the return of that function, you end up in an error because you can only get the reference using the `__getitem__` for example in this way:
```python
from Vector import DynamicVector
from String import String

fn conversion_to_utf(my_string: String) -> DynamicVector[SI8]:
    let vec : DynamicVector[SI8] = my_string.buffer
    return vec
    
print(conversion_to_utf("hello world")[0])
```
And it return the reference to the `0` element of the String, but not the value. 
You can have a look on these issues in which is better explained :)
-> https://github.com/modularml/mojo/issues/126
-> https://github.com/modularml/mojo/issues/113

**Comment by d3v3us** [comment]
hey @czheo, how do opposite, convert DynamicVector[SI8] into String?

**Reply by czheo** [reply]
```py
let s : String = "hello world"
let vec : DynamicVector[SI8] = s.buffer
let p = String(vec.data, len(vec))
print(p) # hello world
```

**Reply by d3v3us** [reply]
thank you @czheo , is there a way to get a type name of given value, typeof is not working)

**Reply by czheo** [reply]
@deveusss I'm not aware of

**Reply by sa-** [reply]
@deveusss This is hacky but you can try to call a non-existent function on the value and the compiler would say `type xzy doesn't implement the function`

**Reply by d3v3us** [reply]
thx, but trying to write more Mojo stype way of doing that

**Reply by d3v3us** [reply]
hey @czheo , not working
```
from Vector import DynamicVector
from String import String
fn into_utf8(data:String) -> DynamicVector[SI8]:
        let vec:DynamicVector[SI8] = data.buffer
        return vec
fn from_utf8(vec:DynamicVector[SI8]) -> String:
        let p = String(vec.data, len(vec))
        return p

let bt=into_utf8(String("temp"))

print(from_utf8(bt))
```

error: Execution was interrupted, reason: signal SIGABRT.
The process has been left at the point where it was interrupted, use "thread return -x" to return to the state before expression evaluation.

What I have missed?

**Reply by mfranzon** [reply]
From my understanding using `fn` you should declare the string as immutable object:
```python
from Vector import DynamicVector
from String import String
fn into_utf8(data:String) -> DynamicVector[SI8]:
        let vec:DynamicVector[SI8] = data.buffer
        return vec
fn from_utf8(vec:DynamicVector[SI8]) -> String:
        let p = String(vec.data, len(vec))
        return p

let s : String = "temp"
let bt=into_utf8(s)
print(from_utf8(bt))
```

**Reply by d3v3us** [reply]
thank you @mfranzon, works like a charm)

**Reply by d3v3us** [reply]
hey @czheo @mfranzon, do you know whats the difference between Pointer.alloc and stack_allocation?

**Reply by sa-** [reply]
I'm not either of them, but I would Pointer.alloc allocates memory on the heap, and stack_allocation does it on the stack as the name suggests

**Reply by d3v3us** [reply]
make sense, my bad, just curious how to use memory.stack_allocation)

**Comment by iamtimdavis** [comment]
Closing this out given there is a clear answer above

**Comment by talhatahir-10xe** [comment]
Can you please share how you wrote SI8?
